stl/inc/ximmortalize.h

```diff
@@ -0,0 +1,118 @@
+// ximmortalize.h internal header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _XIMMORTALIZE_H
+#define _XIMMORTALIZE_H
+#include <yvals.h>
+#ifndef _M_CEE_PURE
+#include <atomic>
+#endif
+
+#if _STL_COMPILER_PREPROCESSOR
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+#ifdef _M_CEE_PURE
+template <class _Ty>
+union _Immortalizer_impl { // constructs _Ty, never destroys
+    _Immortalizer_impl() noexcept : _Storage{} {}
+    _Immortalizer_impl(const _Immortalizer_impl&) = delete;
+    _Immortalizer_impl& operator=(const _Immortalizer_impl&) = delete;
+    ~_Immortalizer_impl() {
+        // do nothing
+    }
+
+    _Ty _Storage;
+};
+
+#pragma warning(push)
+#pragma warning(disable : 4640) // construction of local static object is not thread-safe (/Wall)
+template <class _Ty>
+constexpr _Ty& _Immortalize() noexcept { // return a reference to an object that will live forever
+    /* MAGIC */ static _Immortalizer_impl<_Ty> _Static;
+    return _Static._Storage;
+}
+#pragma warning(pop)
+
+template <class _Ty>
+_Ty& _Immortalize_pointer() noexcept {
+    return _Immortalize<_Ty>();
+}
+
+template <class _Ty>
+_Ty& _Immortalize_two_pointers() noexcept {
+    return _Immortalize<_Ty>();
+}
+
+#else // ^^^ _M_CEE_PURE ^^^ // vvv !_M_CEE_PURE vvv
+template <class _Ty>
+union _Immortalizer_impl {
+public:
+    constexpr _Immortalizer_impl() : _Storage{} {}
+    _Immortalizer_impl(const _Immortalizer_impl&) = delete;
+    _Immortalizer_impl& operator=(const _Immortalizer_impl&) = delete;
+    ~_Immortalizer_impl() {
+        // do nothing
+    }
+
+    _Ty _Storage;
+
+    static _Immortalizer_impl _Instance;
+};
+
+template <class _Ty>
+#ifdef __clang__
+[[clang::require_constant_initialization]]
+#endif // __clang__
+_Immortalizer_impl<_Ty>
+    _Immortalizer_impl<_Ty>::_Instance;
+
+template <class _Ty>
+constexpr _Ty& _Immortalize() noexcept { // return a reference to an object that will live forever
+    return _Immortalizer_impl<_Ty>::_Instance._Storage;
+}
+
+template <class _Ty>
+_Ty& _Immortalize_pointer() noexcept {
+    static atomic<uintptr_t> _Storage;
+    static_assert(sizeof(_Storage) == sizeof(_Ty), "Bad _Immortalize_pointer storage size");

```

Comment:
Not that it makes much difference, since `atomic<uintptr_t>` is lock-free and has the same size as `uintptr_t` on our implementation:
```suggestion
    static_assert(sizeof(uintptr_t) == sizeof(_Ty), "Bad _Immortalize_pointer storage size");
```